content/solids/components/voxel-painter.ts

```diff
@@ -0,0 +1,462 @@
+          setPositionFromVoxelIntersection(placingStart, intersection)
+          placingEnd.copy(placingStart)
+          for (let i = 0; i < 3; i++)
+            placingHelpers[i].position.copy(placingStart)
+        }
+        else mouseControlMode = "rotating"
+      }
+    });
+
+    $canvas.on('pointerup', (event: PointerEvent) => {
+      event.preventDefault();
+
+      if (mouseControlMode === "placing")
+      {
+        const intersection = raycaster.intersectObjects(placingHelpers)[0];
+        placingEnd.copy(intersection.point)
+        snapToNearestValidCubeCenterPosition(placingEnd)
+
+        let iComponent = 0
+        if (placingStart.x === placingEnd.x) iComponent = 1
+        if (placingStart.y === placingEnd.y) iComponent = 2
+        if (placingStart.z === placingEnd.z) iComponent = 0
+        let jComponent = (iComponent + 1) % 3
+        let kComponent = (jComponent + 1) % 3
+
+        let iStart = Math.min( placingStart.getComponent(iComponent), placingEnd.getComponent(iComponent) )
+        let iLimit = Math.max( placingStart.getComponent(iComponent), placingEnd.getComponent(iComponent) )
+        let jStart = Math.min( placingStart.getComponent(jComponent), placingEnd.getComponent(jComponent) )
+        let jLimit = Math.max( placingStart.getComponent(jComponent), placingEnd.getComponent(jComponent) )
+
+        // log(placingStart,placingEnd, iStart,iLimit,jStart,jLimit)
+        let potentialNewPosition = new THREE.Vector3()
+        for ( let i = iStart; i <= iLimit; i+= 1.)
+          for ( let j = jStart; j <= jLimit; j += 1.)
+          {
+            let k = placingStart.getComponent(kComponent)
+            potentialNewPosition.setComponent(kComponent, k)
+            potentialNewPosition.setComponent(iComponent, i)
+            potentialNewPosition.setComponent(jComponent, j)
+            snapToNearestValidCubeCenterPosition(potentialNewPosition)
+            let alreadyOccupied = false
+            voxels.forEach((voxel)=>{
+              if(voxel.position.equals(potentialNewPosition)) alreadyOccupied = true
+            })
+            if(!alreadyOccupied)
+              Voxel().position.copy(potentialNewPosition)
+          }
+      }
+      
+      mouseControlMode = ""
+      placementVisualizer.scale.setScalar(1.)
+    });
+
+    function getCameraSpaceDirectionInAxisPlane(worldSpaceDirection, axisPlane, target)
+    {
+      target.copy(worldSpaceDirection)
+      camera.worldToLocal(target)
+      target[axisPlane] = 0.
+      target.normalize()
+    }
+
+    function getStepTowardDestination(currentValue,destination)
+    {
+      let distanceFromDestination = destination - currentValue
+      let sign = distanceFromDestination == 0. ? 0. : distanceFromDestination / Math.abs(distanceFromDestination)
+      let speed = .01
+      if(speed > Math.abs(distanceFromDestination))
+        speed = Math.abs(distanceFromDestination)
+      return sign * speed
+    }
+
+    updateFunctions.push(()=>
```

Comment:
see above – this can just be a single function called on update

Reply:
Done